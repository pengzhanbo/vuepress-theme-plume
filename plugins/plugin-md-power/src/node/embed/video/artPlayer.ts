/**
 * @[artPlayer](url)
 * @[artPlayer muted autoplay autoMini loop volume=1 poster="xxxx"](url)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { ArtPlayerTokenMeta } from '../../../shared/index.js'
import { isPackageExists } from 'local-pkg'
import { colors } from 'vuepress/utils'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

const installed = {
  dashjs: isPackageExists('dashjs'),
  hlsjs: isPackageExists('hls.js'),
  mpegtsjs: isPackageExists('mpegts.js'),
}

const SUPPORTED_VIDEO_TYPES = ['mp4', 'mp3', 'webm', 'ogg', 'mpd', 'dash', 'm3u8', 'hls', 'ts', 'flv']

export const artPlayerPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<ArtPlayerTokenMeta>(md, {
    type: 'artPlayer',
    name: 'video_artPlayer',
    syntaxPattern: /^@\[artPlayer([^\]]*)\]\(([^)]*)\)/,
    meta([, info, source]) {
      const { attrs } = resolveAttrs<ArtPlayerTokenMeta>(info)
      const url = source.trim()
      checkSupportType(attrs.type ?? url.split('.').pop())

      return {
        autoplay: attrs.autoplay ?? false,
        muted: attrs.muted ?? attrs.autoplay ?? false,
        autoMini: attrs.autoMini ?? false,
        loop: attrs.loop ?? false,
        volume: typeof attrs.volume !== 'undefined' ? Number(attrs.volume) : 0.75,
        poster: attrs.poster,
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '',
        ratio: attrs.ratio ? parseRect(`${attrs.ratio}`) : '',
        type: attrs.type,
        url,
      }
    },
    content({ autoMini, autoplay, loop, muted, poster, url, type, volume, width, height, ratio }) {
      return `<ArtPlayer src="${url}" fullscreen flip playback-rate aspect-ratio setting pip ${
        loop ? ' loop' : ''
      }${
        type ? ` type="${type}"` : ''
      }${
        autoMini ? ' auto-min' : ''
      }${autoplay ? ' autoplay' : ''}${
        muted || autoplay ? ' muted' : ''
      }${
        poster ? ` poster="${poster}"` : ''
      }  :volume="${volume}" width="${width}"${
        height ? ` height="${height}"` : ''
      }${
        ratio ? ` ratio="${ratio}"` : ''
      }/>`
    },
  })
}

function checkSupportType(type?: string) {
  if (!type)
    return

  /* istanbul ignore if -- @preserve */
  if (SUPPORTED_VIDEO_TYPES.includes(type)) {
    let name = ''
    switch (type.toLowerCase()) {
      case 'm3u8':
      case 'hls':
        name = !installed.hlsjs ? 'hls.js' : ''
        break

      case 'flv':
      case 'ts': {
        name = !installed.mpegtsjs ? 'mpegts.js' : ''
        break
      }
      case 'mpd':
      case 'dash':
        name = !installed.dashjs ? 'dashjs' : ''
        break
    }
    /* istanbul ignore if -- @preserve */
    if (name) {
      console.warn(`${colors.yellow('[vuepress-plugin-md-power] artPlayer: ')} ${colors.cyan(name)} is not installed, please install it via npm or yarn or pnpm`)
    }
  }
  else {
    /* istanbul ignore next -- @preserve */
    console.warn(`${colors.yellow('[vuepress-plugin-md-power] artPlayer: ')} unsupported video type: ${colors.cyan(type)}`)
  }
}
