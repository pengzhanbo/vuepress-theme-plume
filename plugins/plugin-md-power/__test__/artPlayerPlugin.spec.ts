import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { artPlayerPlugin } from '../src/node/embed/video/artPlayer.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(artPlayerPlugin)
}

describe('artPlayerPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[artPlayer](/xxx.mp4)

@[artPlayer loop autoplay](/xxx.m3u8)

@[artPlayer autoplay muted volume="0.55"](/xxx.flv)

@[artPlayer width="100%" height="600px" ratio="16:9" auto-mini poster="xx.jpg"](/xxx.mpd)

@[artPlayer type="mp3"](/xxx)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[artPlayer]()

@[artPlayer]xxx

@[ artPlayer]123456

@[ artPlayer](/xxx.mp4)

@[artPlayer type="xxx"](/xxx.xxx)
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
