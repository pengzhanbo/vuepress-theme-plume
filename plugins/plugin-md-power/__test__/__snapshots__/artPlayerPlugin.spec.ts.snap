// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`artPlayerPlugin > should not work 1`] = `
"<ArtPlayer src="" fullscreen flip playback-rate aspect-ratio setting pip   :volume="0.75" width="100%"/><p>@[artPlayer]xxx</p>
<p>@[ artPlayer]123456</p>
<p>@<a href="/xxx.mp4"> artPlayer</a></p>
<ArtPlayer src="/xxx.xxx" fullscreen flip playback-rate aspect-ratio setting pip  type="xxx"  :volume="0.75" width="100%"/>"
`;

exports[`artPlayerPlugin > should work 1`] = `"<ArtPlayer src="/xxx.mp4" fullscreen flip playback-rate aspect-ratio setting pip   :volume="0.75" width="100%"/><ArtPlayer src="/xxx.m3u8" fullscreen flip playback-rate aspect-ratio setting pip  loop autoplay muted  :volume="0.75" width="100%"/><ArtPlayer src="/xxx.flv" fullscreen flip playback-rate aspect-ratio setting pip  autoplay muted  :volume="0.55" width="100%"/><ArtPlayer src="/xxx.mpd" fullscreen flip playback-rate aspect-ratio setting pip  auto-min poster="xx.jpg"  :volume="0.75" width="100%" height="600px" ratio="16:9"/><ArtPlayer src="/xxx" fullscreen flip playback-rate aspect-ratio setting pip  type="mp3"  :volume="0.75" width="100%"/>"`;
